### DISCUSSION TITLE
auto-complete routine signature

### DISCUSSION THREAD

**Discussion by alexksr** [fortran-lang/fortls#395] [Q&A]
**Is your feature request related to a problem? Please describe.**
When calling a function or subroutine with a long list of arguments, there is no autocomplete support, meaning I have to typewrite the arguments names manually reading it from the hover information.

**Describe the solution you'd like**
Parse the list of arguments from the routines signature and autocomplete the dummy argument names.

**Additional context**
I'm using fortls via Modern-Fortran in VScode. I hope this is the right place to ask for help.

Could the auto-completion be improved to provide the routines full signature?

When hovering the function declaration shows up, so the information should be there already. Could the hover-info be parsed to extract the list arguments and auto-complete them?

For example if I have a routine like 
```fortran
subroutine io_save_graph_2d(x, y, f, num_x, num_y, filename, metadata, newlines)
  ... 
end subroutine
```
I want the line "call io_sav" to be auto-completed to
```fortran
call io_save_graph_2d(x=*cursor position*, y=, f=, num_x=, num_y=, filename=, metadata=, newlines=)
```
(Optionally with the cursor set to the indicated position.)

**Comment by gnikit** [comment]
What you are requesting is already present in fortls but it is not exposed in vscode by design, since it makes code very verbose and generally I prefer relying on better variable names and smaller methods rather than very explicit argument parsing interfaces.

To get it to work, you want to _disable_ `--use_signature_help`, which internally sets the `--autocomplete_no_snippets` flag, see [docs](https://fortls.fortran-lang.org/options.html#autocomplete-options). FYI you will notice that for `interfaces` this autocompletion snippet feature is always present.

To do so, you will have to create a JSON configuration file. By default a file with the name `.fortls.json` at the root of your vscode project with the following would override the default and not-exposed fortls settings


```json
{
  "use_signature_help": false
}
```

For more about fortls' file configuration, see the [this docs section](https://fortls.fortran-lang.org/options.html#configuration-using-a-file)

**Comment by gnikit** [comment]
Let me know if this works.

**Comment by alexksr** [comment]
Unfortunately it did not work for me. 
1. The local `.fortls.json` file had no effect even after multiple editor reloads. IDK if this is due to not loading the config file or another reason. How do I access the log data for fortls initialization?
2. I also tried a wrapper script to fortls which I referenced in vscode:
```
#!/bin/bash
fortls --debug_log --config /home/.../bin/myfortls/.fortls.json
```
with 
```
cat .fortls.json 
{
  "use_signature_help": false
}
```
This only resulted in vscode spamming error messages every second upon initialization:
```
[ERROR - 08:51:19] [lsp.client] Unable to verify input arguments with LS:
```

**Comment by gnikit** [comment]
Download the modern Fortran pre release and set the log level to debug.
Then in the output window (where the terminal area is) choose modern
Fortran from the top right dropout menu. There you will see all the stuff
that the extension does including fortls initialization and input arguments.

You might also want to download the dev version of fortls from git with
pip.

**Comment by alexksr** [comment]
I already was on the pre-release branch of Modern Fortran. 

<details> 
  <summary>Here is my log: </summary>

```
[INFO - 12:47:02] Extension Name: Modern Fortran
[INFO - 12:47:02] Extension Version: 3.4.0
[INFO - 12:47:02] Linter set to: "gfortran"
[INFO - 12:47:02] Formatter set to: "findent"
[INFO - 12:47:02] Autocomplete set to: "fortls"
[INFO - 12:47:02] Hover set to: "fortls"
[INFO - 12:47:02] Symbols set to: "fortls"
[INFO - 12:47:02] [lint] Found GNU Fortran version 7.5.0
[DEBUG - 12:47:02] [lint] Using Modern GNU Fortran diagnostics: false
[DEBUG - 12:47:02] [lint] arguments:
[
  "-ffree-line-length-none",
  "-ffixed-line-length-none",
  "-Wall"
]
[DEBUG - 12:47:02] [lint] moduleOutput: -J /home/user/.config/Code/User/workspaceStorage/0c054a5b2019e0cf3cddf76818c09f57/fortran-lang.linter-gfortran/include
[DEBUG - 12:47:02] [lint] Initialising cache for linter.includePaths
[DEBUG - 12:47:02] [lint] glob paths:
[]
[DEBUG - 12:47:02] [lint] resolved paths:
[]
[DEBUG - 12:47:02] [build.single] compiler: "gfortran" located in: "/usr/bin/gfortran"
[INFO - 12:47:02] [build.single] Compiler query command line: /usr/bin/gfortran -fsyntax-only -cpp -ffree-line-length-none -ffixed-line-length-none -Wall -J /home/user/.config/Code/User/workspaceStorage/0c054a5b2019e0cf3cddf76818c09f57/fortran-lang.linter-gfortran/include -ffree-form -o /home/***/repo/src.f90.o /home/***/repo/src.f90
[DEBUG - 12:47:02] [lsp.client] Fortran Language Server -- constructor
[DEBUG - 12:47:02] [lsp.client] Language Server arguments:
[
  "--enable_code_actions",
  "--hover_signature",
  "--use_signature_help",
  "--lowercase_intrinsics",
  "--nthreads=4",
  "--incremental_sync"
]
[DEBUG - 12:47:03] [lsp.client] Language Server version: 1.12.0
[INFO - 12:47:03] [lsp.client] Initialising Language Server for workspace: file:///home/***/repo/
[INFO - 12:47:03] [lsp.client] Language Server arguments: --enable_code_actions --hover_signature --use_signature_help --lowercase_intrinsics --nthreads=4 --incremental_sync
[DEBUG - 12:47:03] [build.single] Compiler output:
/home/***/repo/src.f90:2:4:

 use m_framework
    1
Fatal Error: Can't open module file 'm_framework.mod' for reading at (1): No such file or directory
compilation terminated.

[DEBUG - 12:47:03] [lint] Parsing output to vscode.Diagnostics
[
  {
    "severity": "Error",
    "message": "Can't open module file 'm_framework.mod' for reading at (1): No such file or directory",
    "range": [
      {
        "line": 1,
        "character": 4
      },
      {
        "line": 1,
        "character": 4
      }
    ]
  }
]
[DEBUG - 12:47:03] [lint] Removing duplicates. vscode.Diagnostics are now:
[
  {
    "severity": "Error",
    "message": "Can't open module file 'm_framework.mod' for reading at (1): No such file or directory",
    "range": [
      {
        "line": 1,
        "character": 4
      },
      {
        "line": 1,
        "character": 4
      }
    ]
  }
]
```
</details>

There is no config file loaded. The passed arguments seem to be the default ones.

Sorry for having to ask: Where can I get the dev version?
